src/Microsoft.OData.Core/UriParser/Resolver/ODataUriResolver.cs

```diff
@@ -378,19 +378,34 @@ public virtual IEnumerable<IEdmOperationImport> ResolveOperationImports(IEdmMode
         /// <returns>The resolved key list.</returns>
         public virtual IEnumerable<KeyValuePair<string, object>> ResolveKeys(IEdmEntityType type, IDictionary<string, string> namedValues, Func<IEdmTypeReference, string, object> convertFunc)
         {
+            if (!TryResolveKeys(type, namedValues, convertFunc, out IEnumerable<KeyValuePair<string, object>> resolvedKeys))
+            {
+                throw ExceptionUtil.CreateBadRequestError(Strings.BadRequest_KeyMismatch(type.FullName()));
+            }
+
+            return resolvedKeys;
+        }
+
+        /// <summary>
+        /// Attempts to resolve keys for a certain entity set.
+        /// This function would be called when key is specified as name value pairs. E.g. EntitySet(ID='key')
+        /// </summary>
+        /// <param name="type">The type for the current EntitySet</param>
+        /// <param name="namedValues">The dictionary of name-value pairs.</param>
+        /// <param name="convertFunc">The converter function to be used for value conversion.</param>
+        /// <param name="resolvedKeys">If the resolution was successful, this will contain the list of resolved keys.</param>
+        /// <returns>True if the key resolution was successful, otherwise false.</returns>
+        internal bool TryResolveKeys(IEdmEntityType type, IDictionary<string, string> namedValues, Func<IEdmTypeReference, string, object> convertFunc, out IEnumerable<KeyValuePair<string, object>> resolvedKeys)
+        {
+            resolvedKeys = null;
             var convertedPairs = new Dictionary<string, object>(StringComparer.Ordinal);
 
-            // Throw an error if key size from url doesn't match that from model.
-            // Other derived ODataUriResolver intended for alternative key resolution, such as the built in AlternateKeysODataUriResolver,
-            // should override this ResolveKeys method.
             IEnumerable<IEdmStructuralProperty> keys = type.Key();
-            if (keys.Count() != namedValues.Count)
-            {
-                throw ExceptionUtil.CreateBadRequestError(Strings.BadRequest_KeyCountMismatch(type.FullName()));

```

Comment:
Can we conclude that the exception with the `BadRequest_KeyCountMismatch` message will no longer be raised?

Reply:
Yeah, you're right. I can remove it.

Reply:
Actually it's still used. There's another `ResolveKeys()` overload that deals with positional keys (as opposed to named keys) that still uses this error message.